### ISSUE TITLE
ev-click event binding not working, what dependencies do I need? (dom-delegator? value-event?)

### ISSUE THREAD

**Issue by th3james** [Matt-Esch/virtual-dom#166]
I'm trying to build my own lightweight MVI system using `virtual-dom` in my view layer, but I'm struggling to get event binding to work.

The virtual-hyperscript README suggests that I should be able to bind events using `ev-click`, but none of my bound functions get fired:
https://github.com/Matt-Esch/virtual-dom/tree/master/virtual-hyperscript

```javascript
var virtualEl = virtualDom.h('div', {'ev-click': function(ev) {alert("Yo");}}, ['clickable']);
var domEl = virtualDom.create(virtualEl);
$('body').append(domEl);
// clicking created element has no effect
```

Looking around [this](https://github.com/Matt-Esch/virtual-dom/wiki#events) and the mercury source suggests that I might need some extra modules like `dom-delegator` and `value-event` to get event binding working, but neither of the README's explain how to make it work with the `ev-click` syntax.

I should add that I'm not using browserify. Thus far I've built and included `virtual-dom` and `dom-delegator` by downloading them and using `browserify --standalone` on each repo. I've included them both in my project, and created a delegator before creating elements using `virtual-hyperscript`:
```javascript
// before virtual-dom element creation (but after source include)
delegator = new DOMDelegator()
```
I've pulled this all together in a small github repo to make it easier to see what I'm doing:
https://github.com/th3james/Virtual-dom-event-binding-issue/blob/master/index.html

I can't quote follow how `dom-delegator` sees and binds to the events? My understanding from this issue is it uses a shared map:
https://github.com/Raynos/mercury/issues/93
...but looking at the source, I can't see how this connection is made.

Nudges in the right direction greatly appreciated, as I'm really keen to use `virtual-dom`.

**Comment by staltz** [comment]
Use `onclick: function (ev) { ...`

Use dom-delegator only if you know what you're doing.

**Comment by th3james** [comment]
Ah, that's a simple fix! Thanks for the response.

Might I suggest an adjustment to the README for [virtual-hyperscript](https://github.com/Matt-Esch/virtual-dom/tree/master/virtual-hyperscript) to make it clear that `ev-event` and `dom-delegator` aren't necessary? Given that that's the only mention of events, it does rather imply that `ev-event` is the way to do it.

**Comment by Raynos** [comment]
@th3james 

Good point we should update the documentation on virtual-hyperscript to mention `onclick` etc events.

As for why `dom-delegator` and `virtual-dom` dont work together, `virtual-dom@1.0.0` is a breaking change and we havn't updated dom-delegator yet. mercury still uses an older version of virtual-dom.

**Comment by deepak1556** [comment]
with this change https://github.com/Raynos/dom-delegator/pull/18 , i am able use `dom-delegator` with `virtual-dom` https://gist.github.com/deepak1556/465492d9319e530534b2 :)

**Comment by matthewp** [comment]
@Raynos Mercury is using the latest virtual-dom now, was dom-delegator updated as well, or....? My ev-clicks are no longer working.

**Comment by Raynos** [comment]
@matthewp latest version of mercury uses the correct dom-delegator & virtual-dom combinations. You can verify with `npm ls dom-delegator`

**Comment by matthewp** [comment]
I see that I have the latest. Has the api changed? My ev-click functions are being set as the string value to the html attribute "ev-click" which is i'm guessing not as intended.

**Comment by Raynos** [comment]
@matthewp Are we talking server side or client side ?

Client side this code should trigger ( https://github.com/Matt-Esch/virtual-dom/blob/master/virtual-hyperscript/index.js#L94-L97 ). Server side the stringify function might accidentally serialize hooks which is a stringify bug.

**Comment by matthewp** [comment]
@Raynos Upon further investigation the events are working fine, this is just a stringify bug.

**Comment by kuraga** [comment]
Which lines of `virtual-dom`'s code do use `dom-delegator`? I can't find `elegat` at all...

**Comment by neonstalwart** [comment]
@kuraga virtual-dom does not use dom-delegator.  virtual-dom and dom-delegator both use a shared copy of ev-store though which is what raynos/mercury#22 is about

**Comment by jesseditson** [comment]
@deepak1556 trying out your gist in requirebin and I don't see any of the events fire - is that still working or does dom-delegator not work with virtual-dom anymore?

**Comment by deepak1556** [comment]
@jesseditson it works fine, here try this another sample http://requirebin.com/?gist=6a1736bce2c18c09533c

**Comment by jesseditson** [comment]
@deepak1556 that one works!

**Comment by StreetStrider** [comment]
Hello. Please, can anyone tell what's the status on this? It seemes that I'm experiencing this issue. I'm using `virtual-dom@2.0.1` as CommonJS and events not working.

I've read this in the docs:
> Note: You must create an instance of dom-delegator for ev-* to work.

The documentation could be more verbose here. Should I kickstart event storage manually? Or if I'm using CommonJS modules, event storage is constructed for me automatically (thanks to dependencies)? Should be anything putted to my `package.json` manifest beside `virtual-dom`? How to start event storage properly? etc

**Comment by iambumblehead** [comment]
It would be great if the documentation described dom-delegator delegation. Does a list of links using ev-click result in many event handlers? Or does a single handler delegate events to the ev-click functions?

**Comment by neonstalwart** [comment]
dom-delegator has one handler for each type of event. this is the list of
events (
https://github.com/Raynos/dom-delegator/blob/f0ed1a0d137846850653dbdcf906746fb1bc1b38/index.js#L13).
 events are listened to in the capture phase (
https://github.com/Raynos/dom-delegator/blob/f0ed1a0d137846850653dbdcf906746fb1bc1b38/dom-delegator.js#L84)
so that events like blur and focus which don't bubble can still be handled
with a single listener.

On Tue, Dec 29, 2015 at 2:37 PM iambumblehead <notifications@github.com>
wrote:

> It would be great if the documentation described dom-delegator delegation.
> Does a list of links using ev-click result in many event handlers? Or does
> a single handler delegate events to the ev-click functions?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Matt-Esch/virtual-dom/issues/166#issuecomment-167883080>
> .
>
-- 

--
ben...

**Comment by iambumblehead** [comment]
@neonstalwart thank you

**Comment by iambumblehead** [comment]
an additional point --each time the vnode is published to the document, the number of listeners returned in chrome matches the number found in the most defined in the vnode.

```javascript
  var tree = (
    h('ul#mylistff', [
      h('li', { onclick : e => console.log(e) }),
      h('li', { onclick : e => console.log(e) })
    ]));

  var newTree = (
    h('ul#mylistff', [
      h('li', { onclick : e => console.log(e) }),
      h('li', { onclick : e => console.log(e) }),
      h('li', { onclick : e => console.log(e) })      
    ]));

  var newnewTree = (
    h('ul#mylistff', [
      h('li', { onclick : e => console.log(e) })      
    ]));
  
  var treeelem = createElement(tree);
  document.body.appendChild(treeelem);
  
  var patches = diff(tree, newTree);
  var rootNode = patch(document.getElementById('mylistff'), patches);

  patches = diff(newTree, newnewTree);
  rootNode = patch(document.getElementById('mylistff'), patches);

  console.log('events: ', getevents(document.getElementById('mylistff')).length); // "events: 1"
```
